### ISSUE TITLE
Try out AtomicTree abstraction for storage layer

### ISSUE THREAD

**Issue by michielbdejong** [michielbdejong/ldp-resource-store#2]
Slight variation on @RubenVerborgh's ResourceStore, with 1) adding distinction between containers and resources as per https://github.com/RubenVerborgh/solid-server-ts/issues/6#issuecomment-478749248, 2) data in resources as `any` (no understanding of RDF vs non-RDF at this level, feels a bit cleaner to me than Ruben's https://github.com/RubenVerborgh/solid-server-ts/blob/master/src/ldp/Patch.ts proposal), and then use read/write locks to still allow atomic PATCH updates and glob reads.

I'm not fully certain if the path should be a string with `/` separator, so basically the normalized URL path, but I think so.

I'll see what the code looks like with this. At this point it's still a small change, and I think an improvement on current abstraction:

- getWriteLock(path) → lock.release method + WritableContainer or WritableResource or WritableEmptySpace
  - empty:
    - createContainer(path) like mkdir -p
    - createResource(path) like mkdir -p; creates empty Resource
  - container:
    - delete (also deletes contained containers and resources)
  - resource:
    - getData (needed for implementing atomic PATCH updates)
    - writeData
    - delete
- getReadLock(path) -> lock.release method + + ReadableContainer or ReadableResource or ReadableEmptySpace
  - container:
    - getMembers
  - resource:
    - getData
  - empty:
    (nothing)
- on(‘change’)

**Comment by RubenVerborgh** [comment]
Preliminary comment: leave locks out of the interface. The operations are atomic already (on purpose); locking can be achieved by a decorator pattern (dedicated lock) or by individual implementations. (Same for quota.)

**Comment by michielbdejong** [comment]
Thanks! I can see why you left locking out of your ResourceStore, because it understands Patch, but I wanted to try it this way and see what the result looks like.

Apart from PATCH, locking is needed for when you stream a long Resource that gets changed when you're half way through reading.

**Comment by RubenVerborgh** [comment]
Agree on need for locking, and definitely useful to compare consequences of locking in or out the interface. Would need to see more code probably (cfr. my example) for meaningful input. Good thinking.

**Comment by michielbdejong** [comment]
What I ended up with is:
* AtomicTree
  * getReadLockedContainer(path)
    * exists?
    * getMembers => Array<string>
    * releaseLock
  * getReadLockedResource(path)
    * exists?
    * getData -> { contentType, body } 
    * releaseLock
  * getReadWriteLockedContainer(path), same as getReadLockedContainer plus:
    * delete
    * reset (creates an empty container where none exists)
  * getReadWriteLockedResource(path), same as getReadLockedResource plus:
    * delete
    * reset (creates an empty resource where none exists)
    * setData

PATCH can then get a read-write locked resource for the path, see what it needs to do, and release the lock when done. That way, I was able to get rid of the Works quite nicely so far!

**Comment by RubenVerborgh** [comment]
My gut feeling is that a interface with locks as the primary abstraction will make some things harder to build. Let's say something like quota, for instance.

But we probably should see both alternatives in a larger architectural picture; it's hard to make decisions only on the micro-level.

**Comment by michielbdejong** [comment]
Yes, we'll see! :) I'm also thinking about experimenting with microservices-like code architecture, in the sense of making task queues: parse LDP, determine identity, determine ACL, determine trusted apps, determine allowed modes, then split to: container read, glob read, resource read, container add, quota check, resource write, resource update, container delete, resource delete, and then all coming back to a single respond and release.

**Comment by michielbdejong** [comment]
@RubenVerborgh the AtomicTree interface is an abstraction of only the storage backend, it doesn't make any quota decisions, although for the quota check to be possible, it will probably need to get a function that reports current storage usage. I now refactored the business logic code to have workers that pass on tasks to each other. The decision of whether a request needs to go via the QuotaChecker or not is actually a good example, it happens [after the determineAllowedModes step, before the fork to the various LdpTasks](https://github.com/michielbdejong/ldp-resource-store/blob/b13069f28feb0da4b2ac3fb9d0cd291ea9d01afe/src/ldp.ts), so it's the [AclChecker worker](https://github.com/michielbdejong/ldp-resource-store/blob/b13069f28feb0da4b2ac3fb9d0cd291ea9d01afe/src/workers/AclChecker.ts#L42-L45) that makes that decision, based on information that the [LdpParser worker](https://github.com/michielbdejong/ldp-resource-store/blob/b13069f28feb0da4b2ac3fb9d0cd291ea9d01afe/src/workers/LdpParser.ts#L30-L32) put on in the LdpParserResult.

**Comment by RubenVerborgh** [comment]
> although for the quota check to be possible, it will probably need to get a function that reports current storage usage

And that's my problem with it. A good test for the interface is to think of various future extension points (such as quota) to see whether we can achieve them without having to change the interface.

> so it's the [AclChecker worker](https://github.com/michielbdejong/ldp-resource-store/blob/b13069f28feb0da4b2ac3fb9d0cd291ea9d01afe/src/workers/AclChecker.ts#L42-L45) that makes that decision,

The AclChecker should not know/care about what the next step is; that should be orchestrated on a higher level. A currently written, this requires AclChecker to have knowledge about quota checking, which should be completely independent.

Perhaps to be discussed in a separate thread, but what is the necessity for the task/workers architecture?

**Comment by michielbdejong** [comment]
yeah, i was thinking later that it's probably nicer to throw an error from the storage layer when quota are exceeded. you're right though that for that to work, we need to define that `ReadWriteLockedContainer#addMember` and `ReadWriteLockedResource#setData` can throw a `QuotaExceededError`. I don't see how any storage layer interface could pass your test of being able to add a quota check without changing the interface?

Regarding higher layer for the task routing, yes, each worker should just post results to queues that are labeled in the context of the result of the current step, not in terms of what the next step will be. I'll fix that.

**Comment by pmcb55** [comment]
Hi guys, just seeing this now, and I too would have very serious concerns about exposing locking semantics in an abstract interface. Without delving into this too much, I would have thought a mechanism like Memento might be able to handle this better (i.e. Michiel's example of streaming a changing resource reverts to simply streaming a fixed 'version' of that resource (where each atomic change to a resource results in a new version of that resource)).
I'd also have concerns about manually wiring up (using Worker threads/task queues) what sounds like complex workflow (e.g. "**parse LDP, determine identity, determine ACL, determine trusted apps, determine allowed modes, then split to: container read, glob read, resource read, container add, quota check, resource write, resource update, container delete, resource delete, and then all coming back to a single respond and release**"). That all sounds very much like a need for proper Enterprise Integration Patterns (EIP), which for Java I'd propose Apache Camel (which Trellis already uses internally for example). (But I don't know of any Camel-like EIP implementation in Javascript, but then I haven't looked either!)

**Comment by RubenVerborgh** [comment]
> I don't see how any storage layer interface could pass your test of being able to add a quota check without changing the interface?

With https://github.com/RubenVerborgh/solid-server-ts/blob/master/src/ldp/ResourceStore.ts, the idea was to use the decoration pattern, such that there would be an implementation `QuotaChecker implements IResourceStore` that would perform the checks on `addResource`, `setRepresentation`, and `modifyResource`.

**Comment by michielbdejong** [comment]
> there would be an implementation QuotaChecker implements IResourceStore

but then you would still be changing the interface because you would be introducing `throws QuotaExceptionError`, right?

**Comment by RubenVerborgh** [comment]
They're throws anything as far as I'm concerned.

**Comment by michielbdejong** [comment]
ah ok, then we're good.

**Comment by michielbdejong** [comment]
@pmcb55 Memento is out of scope, but i thought adding the releaseLock() function was a nice touch towards making the interface actually theoretically correct, since in theory, we shouldn't ignore race conditions. Having said that, if storage-implementers don't care about locking then they can just fake it, and you're basically back at what NSS does on the filesystem backend.

Re request flow, I separated out the flow control from the worker implementations, agree with both of you that that's important to keep separate.

#4 is an attempt to make the code architecture more microservices-like without actually introducing queues and serialization. It's a balancing act, I'm trying very hard to satisfy all our design philosophies at the same time here, while producing working code in the process :)

**Comment by michielbdejong** [comment]
Merged this into #1 now.

**Comment by pmcb55** [comment]
@michielbdejong I know you've closed this issue, but just for future reference, I'm not sure what you meant by 'Memento is out of scope', since it's just a standard design pattern (straight from the original GoF book): https://en.wikipedia.org/wiki/Memento_pattern
So my suggestion above was just to think about using one of the most standard design patterns to address the architectural design/abstract interface for your problem (i.e. atomicity).

**Comment by michielbdejong** [comment]
Ah sorry I thought you meant exposing [Memento, the protocol](https://tools.ietf.org/html/rfc7089) which a thing that Trellis supports but that's not part of the Solid spec.

Memento as design pattern could be useful! Let's discuss today.
